src/controller/controller.c

```diff
@@ -440,6 +477,69 @@ static bool controller_setup_node_connection_handler(Controller *controller) {
         return true;
 }
 
+static int controller_reset_heartbeat_timer(Controller *controller, sd_event_source **event_source);
+
+static int controller_heartbeat_timer_callback(sd_event_source *event_source, UNUSED uint64_t usec, void *userdata) {
+        Controller *controller = (Controller *) userdata;
+        Node *node = NULL;
+        int r = 0;
+
+        LIST_FOREACH(nodes, node, controller->nodes) {
+                struct timespec now;
+                r = clock_gettime(CLOCK_REALTIME, &now);
+                if (r < 0) {
+                        return r;
+                }
+                time_t dsec = now.tv_sec - node->last_seen.tv_sec;
+                if (dsec > controller->heartbeat_threshold_msec / 1000 ||
```

Comment:
Please review `time-util.h` and `time-util.c` files, there are some constants and util functions for stuff like this. It is better to use them or add more if needed.